Chapter 10 – Debugging
https://automatetheboringstuff.com/chapter10/

# デバッグ

より複雑なプログラムを書くのに十分な知識があるので、単純ではないバグを見つけることができます。   
この章では、プログラム内のバグの根本的な原因を突き止めるためのツールやテクニックについて説明します。  

プログラマーの間で古いジョークを言い換えると、「コードを書くことはプログラミングの90％を占めます。 デバッグコードは他の90％を占めています。」  

あなたのコンピュータはあなたの言うことだけを行います。   
あなたの心を読んで、あなたが意図したことをすることはありません。   
プロのプログラマーでもバグが常に発生するので、プログラムに問題がある場合は落胆しないでください。  

幸いにも、コードが正確に何をしているのか、どこが間違っているのかを特定するためのいくつかのツールとテクニックがあります。   
まず、ロギングとアサーションを見ていきます。  
これは、バグを早期に発見するのに役立つ2つの機能です。   
一般的に、バグを早く発見すればするほど修正が容易になります。

次に、デバッガの使い方を見ていきます。   
デバッガはIDLEの機能であり、一度に1命令ずつプログラムを実行するので、コード実行中に変数の値を検査したり、プログラムの途中で値がどのように変化するかを追跡したりすることができます。   
プログラムをフルスピードで実行するよりもはるかに時間がかかりますが、実行中のプログラムの実際の値をソースコードから推測するのではなく、実際の値を調べると便利です。

# 例外の発生

Pythonは、無効なコードを実行しようとするたびに例外を発生させます。   
第3章では、tryやexcept文でPythonの例外を処理する方法について説明しました。  
これにより、プログラムは予期した例外から回復することができます。   
しかし、コード内で独自の例外を発生させることもできます。   
例外を発生させる方法は、「この関数でコードを実行しないで、プログラムの実行をexceptステートメントに移す」という方法です。

raise文で例外が発生します。 コードでは、raise文は次の要素で構成されます。
- raiseキーワード
- exception()関数の呼び出し
- exception()関数に渡される有用なエラーメッセージを含む文字列

```python
>>> raise Exception('This is the error message.')
Traceback (most recent call last):
  File "<pyshell#191>", line 1, in <module>
    raise Exception('This is the error message.')
Exception: This is the error message.
```

tryおよびexceptステートメントがない場合、raiseステートメントで例外を発生させるとプログラムは単にクラッシュし、例外のエラーメッセージを表示します。  

多くの場合、関数を呼び出すコードであり、関数自体ではなく、expectionを処理する方法を知っています。   
したがって、関数内でraise文が一般的に表示され、関数を呼び出すコード内でtry文とexcept文が表示されます。

ここでは、文字、幅、高さを取るboxPrint（）関数を定義し、その幅と高さの箱を少し描写するために文字を使用します。   
このボックス形状はスクリーンに印刷されます。  

文字を1文字にし、幅と高さを2より大きくしたいとしましょう。  
これらの要件が満たされない場合、例外を発生させるif文を追加します。   
後で、さまざまな引数を指定してboxPrint（）を呼び出すと、try / exceptは無効な引数を処理します。

このプログラムは、exceptステートメントの例外としてerr以外の例外を使用します。   
例外オブジェクトがboxPrint（）から返された場合、このexcept文はerrという名前の変数に格納します。   
次に、Exceptionオブジェクトをstr（）に渡して文字列に変換し、ユーザーフレンドリーなエラーメッセージを生成することができます。   
このboxPrint.pyを実行すると、出力は次のようになります。

```
****
*  *
*  *
****
OOOOOOOOOOOOOOOOOOOO
O                  O
O                  O
O                  O
OOOOOOOOOOOOOOOOOOOO
An exception happened: Width must be greater than 2.
An exception happened: Symbol must be a single character string.
```

tryおよびexceptステートメントを使用すると、プログラム全体がクラッシュするのではなく、エラーをより適切に処理できます。
